### PR TITLE
Preparation for Pico Extensibility Enhancements

### PR DESCRIPTION
**Pull request by trentjeff**
This resolves issue #6891

### CODE REVIEW

File: pico/api/src/main/java/io/helidon/pico/api/ServiceProviderBindable.java [hunk 1]
```diff
@@ -108,11 +108,20 @@ default void rootProvider(ServiceProvider<T> rootProvider) {
         throw new UnsupportedOperationException();
     }
 
+    /**
+     * Returns the previously assigned {@link PicoServices} instance.
+     *
+     * @return the previously assigned pico services instance, or empty if never assigned
+     * @see #picoServices(Optional)
+     */
+    Optional<PicoServices> safePicoServices();
```
Comment (tomas-langer) [inline pico/api/src/main/java/io/helidon/pico/api/ServiceProviderBindable.java:117]: Should be called `picoServices`, this name is strange.

Comment (tomas-langer) [reply pico/api/src/main/java/io/helidon/pico/api/ServiceProviderBindable.java:117]: There is a conflict with `picoServices` on AbstractProvider - that method should be removed, and this one used instead (with `orElseThrow()` maybe)

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: pico/api/src/main/java/io/helidon/pico/api/ServiceProviderBindable.java (hunk 1 above)
Comment (trentjeff) [reply pico/api/src/main/java/io/helidon/pico/api/ServiceProviderBindable.java]: Addressed and pushed.

File: pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java [hunk 2]
```diff
@@ -202,10 +202,16 @@ public Phase currentActivationPhase() {
...
      * Used to access the current pico services instance assigned to this service provider.
      *
+     * @param expected true to indicate that the return values is expected to be non-empty
      * @return the pico services assigned to this service provider
      */
-    public PicoServices picoServices() {
-        return Objects.requireNonNull(picoServices, description() + ": picoServices should have been previously set");
+    @Override
+    public Optional<PicoServices> picoServices(boolean expected) {
+        if (expected) {
```
Comment (tomas-langer) [inline pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java:210]: The whole idea of return an optional is to leave the decision whether it is expected or not on the caller of the method.
This method should just `return Optional.ofNullable(picoServices)`
The caller should throw an exception if they should be available

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java (hunk 2 above)
Comment (trentjeff) [reply pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java:210]: I wanted the provider to control the exception and not the caller. I'll address it in a slightly modified way.

Comment (trentjeff) [reply pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java]: pushed

File: pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java [hunk 3]
```diff
@@ -199,13 +199,17 @@ public Phase currentActivationPhase() {
...
-     */
-    public PicoServices picoServices() {
-        return Objects.requireNonNull(picoServices, description() + ": picoServices should have been previously set");
+    @Override
+    public Optional<PicoServices> picoServices() {
+        return Optional.ofNullable(picoServices);
+    }
+
+    PicoServices requiredPicoServices() {
+        PicoServices picoServices = picoServices().orElse(null);
```
Comment (tomas-langer) [inline pico/runtime/src/main/java/io/helidon/pico/runtime/AbstractServiceProvider.java:208]: you can do
```
PicoServices picoServices = picoServices()
     .orElseThrow(() -> new PicoServiceProviderException(description() + ": picoServices should have been previously set", this));
```

**Review by tomas-langer — APPROVED** [review]

**Pull request merged**